flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/statistics/ContinuousHistogram.scala

```diff
@@ -0,0 +1,315 @@
+        val m: Int = bins
+        val n: Int = temp.bins
+        var i, j: Int = 0
+        val mergeList = new mutable.ArrayBuffer[(Double, Int)]()
+        while (i < m || j < n) {
+          if (i >= m) {
+            mergeList += ((temp.getValue(j), temp.getCounter(j)))
+            j = j + 1
+          } else if (j >= n || getValue(i) <= temp.getValue(j)) {
+            mergeList += data.apply(i)
+            i = i + 1
+          } else {
+            mergeList += ((temp.getValue(j), temp.getCounter(j)))
+            j = j + 1
+          }
+        }
+        // the size will be brought to capacity while constructing the new histogram itself
+        val finalLower = Math.min(lower, temp.lower)
+        val finalUpper = Math.max(upper, temp.upper)
+        val ret = new ContinuousHistogram(B, finalLower, finalUpper)
+        ret.loadData(mergeList.toArray)
+        ret
+      }
+      case default =>
+        throw new RuntimeException("Only a continuous histogram is allowed to be merged with a " +
+          "continuous histogram")
+
+    }
+  }
+
+  /** Returns the qth quantile of the histogram
+    *
+    * @param q Quantile value in (0,1)
+    * @return Value at quantile q
+    */
+  def quantile(q: Double): Double = {
+    require(bins > 0, "Histogram is empty")
+    require(q > 0 && q < 1, "Quantile must be between 0 and 1")
+    val wantedSum = (q * total).round.toInt
+    var currSum = count(getValue(0))
+
+    if (wantedSum < currSum) {
+      require(lower > -MaxValue, "Set a lower bound before proceeding")
+      return Math.sqrt(2 * wantedSum * Math.pow(getValue(0) - lower, 2) / getCounter(0)) + lower
+    }
+
+    /** Walk the histogram to find sums at every bin value
+      * As soon as you hit the interval where you should be
+      * Walk along the trapezoidal line
+      * This leads to solving a quadratic equation.
+      */
+    for (i <- 1 to bins - 1) {
+      val tmpSum = count(getValue(i))
+      if (currSum <= wantedSum && wantedSum < tmpSum) {
+        val neededSum = wantedSum - currSum
+        val a: Double = getCounter(i) - getCounter(i - 1)
+        val b: Double = 2 * getCounter(i - 1)
+        val c: Double = -2 * neededSum
+        if (a == 0) {
+          return getValue(i - 1) + (getValue(i) - getValue(i - 1)) * (-c / b)
+        } else {
+          return getValue(i - 1) +
+            (getValue(i) - getValue(i - 1)) * (-b + Math.sqrt(b * b - 4 * a * c)) / (2 * a)
+        }
+      } else {
+        currSum = tmpSum
+      }
+    }
+    require(upper < MaxValue, "Set an upper bound before proceeding")
+    // this means wantedSum > sum(getValue(bins-1))
+    // this will likely fail to return a bounded value.
+    // Make sure you set some proper limits on min and max.
+    getValue(bins - 1) + Math.sqrt(
+      Math.pow(upper - getValue(bins - 1), 2) * 2 * (wantedSum - currSum) / getCounter(bins - 1))
+  }
+
+  /** Returns the probability (and by extension, the number of points) for the value b
```

Comment:
Rhetorical again. Look at the function you're describing. It's called count and you write something about probability. How does this fit? And I know that it can be seen as the scaled probability. But if you describe things as a transformation of the actual thing you want to describe, then you will only confuse people. E.g. you could also apply the logarithm to the probabiliy value and then talk about a value in the range [-inf, 0]. Isn't this a little bit confusing?